src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,670 @@
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+            if (completionService == null)
+            {
+                var listWithSelections = GetHighlightedList(itemsInList, filterText);
+                return new EditorCompletion.FilteredCompletionModel(listWithSelections.ToImmutableArray(), 0, filters, mustSetSelection ? EditorCompletion.UpdateSelectionHint.SoftSelected : EditorCompletion.UpdateSelectionHint.NoChange,
+                    centerSelection: true, uniqueItem: default);
+            }
+
+            var matchingItems = itemsInList
+                .Where(r => r.MatchedFilterText)
+                .Select(t => GetOrCreateRoslynItem(t.CompletionItem))
+                .AsImmutable();
+
+            var chosenItems = completionService.FilterItems(snapshot.GetOpenDocumentInCurrentContextWithChanges(), matchingItems, filterText);
+
+            var recentItems = _recentItems;
+            var bestItem = GetBestItemBasedOnMRU(chosenItems, recentItems);
+            var highlightedList = GetHighlightedList(itemsInList, filterText).ToImmutableArray();
+            var updatedFilters = GetUpdatedFilters(sortedList, itemsInList, filters, filterText);
+
+            // TODO: Can we get away with less complexity here by only doing hard select on preselection and not on regular filter text matching / etc...
+            // https://github.com/dotnet/roslyn/issues/29108
+            var isHardSelection = IsHardSelection(bestItem, snapshot, caretPosition, filterReason, filterText, triggerReason);
+
+            if (bestItem == null)
+            {
+                return new EditorCompletion.FilteredCompletionModel(highlightedList, 0, updatedFilters, isHardSelection ? EditorCompletion.UpdateSelectionHint.Selected : EditorCompletion.UpdateSelectionHint.SoftSelected, centerSelection: true, uniqueItem: null);
+            }
+
+            // TODO: Better conversion between Roslyn/Editor completion items
+            // https://github.com/dotnet/roslyn/issues/29106
+            var selectedItemIndex = itemsInList.IndexOf(i => i.CompletionItem.DisplayText == bestItem.DisplayText);
+
+            EditorCompletion.CompletionItem uniqueItem = null;
+            if (bestItem != null && matchingItems.Length == 1 && filterText.Length > 0)
+            {
+                uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+            }
+
+            return new EditorCompletion.FilteredCompletionModel(highlightedList, selectedItemIndex, updatedFilters, isHardSelection ? EditorCompletion.UpdateSelectionHint.Selected : EditorCompletion.UpdateSelectionHint.SoftSelected, centerSelection: true, uniqueItem);
+        }
+
+        private RoslynCompletionItem GetOrCreateRoslynItem(EditorCompletion.CompletionItem item)
+        {
+            if (item.Properties.TryGetProperty<RoslynCompletionItem>("RoslynItem", out var roslynItem))
```

Comment:
Constant since I'm guessing this is hardoced in other places?

Reply:
the only other instance is in test state. so, it think constant is not necessary.

---
In reply to: [208430212](https://github.com/dotnet/roslyn/pull/29016#discussion_r208430212) [](ancestors = 208430212)

Reply:
@ivanbasov Still having a constant would mean that's possible to determine.